### PR TITLE
Select Mode: Blocks outside the main sections root should be disabled

### PR DESCRIPTION
**Pull request by youknowriad**
Related #60021 

## What?

I noticed a bug where sometimes in "select mode" some random content blocks remain editable even if they don't belong in the "main" section. For instance in 2024, you can see that you can check the text of some menu items in the footer while editing the home page template in select mode.

This bug led me to do a small refactoring to the `getBlockEditingMode` function, added some comments and also added unit tests to cover all the "navigation mode" use-cases.

## Testing Instructions

1- Open 2024 home page template
2- Pick the "select tool" (or edit mode)
3- Check that you can't edit random text in the footer.

### CODE REVIEW

File: packages/block-editor/src/store/selectors.js [hunk 1]
```diff
@@ -2968,43 +2976,66 @@ export const getBlockEditingMode = createRegistrySelector(
...
+				if ( sectionsClientIds.includes( clientId ) ) {
+					return 'contentOnly';
+				}
+
+				// Blocks outside sections should be disabled.
+				const isWithinSectionRoot = isWithinBlock(
+					state,
+					clientId,
+					sectionRootClientId
+				);
```
Comment (getdave) [inline packages/block-editor/src/store/selectors.js:3001]: Do we also want to do this in `zoom-out` mode? I'd say we might well.

Aside: this is why maintaining x2 similar modes during the 6.7 release cycle is going to be tricky.

Comment (youknowriad) [reply packages/block-editor/src/store/selectors.js:3001]: What if we just change the condition to `if [ "zoom-out", "navigation" ].includes( editorMode )` (until we remove zoom-out)

**Comment by getdave** [comment]
I think Zoom Out has a similar requirement. In https://github.com/WordPress/gutenberg/issues/65145 the reporter is noting that blocks outside the section root aren't editable.

File: packages/block-editor/src/store/selectors.js (hunk 1 above)
Comment (youknowriad) [reply packages/block-editor/src/store/selectors.js:3001]: Here are the two differences that exist with zoom-out today:

 - The root container is "default" in "navigation" but "contentOnly" in zoom-out.
 - Zoom-out doesn't enable "content blocks", everything is disabled within sections. 

Do we want to unify that?

Comment (youknowriad) [reply packages/block-editor/src/store/selectors.js:3001]: Given that currently zoom-out disables everything that is not a section, the bug covered by this PR is actually already covered by "zoom-out" but if we change zoom-out to enable editing "content " blocks, then we'll have the same issue. So for now, it seems safe to move forward.

**Review by draganescu — APPROVED** [review]
This fixes the problem and the comments and tests are great!

File: packages/block-editor/src/store/selectors.js (hunk 1 above)
Comment (getdave) [reply packages/block-editor/src/store/selectors.js:3001]: Thank you 🙇‍♂️

**Comment by youknowriad** [comment]
I think this PR caused a negative impact on the "typing" performance metrics. Around 5% to 10%. This is most likely due to the increased complexity of the `getBlockEditingMode` selector although, that's not a certainty.

**Pull request merged**